libgeos/libgeos_handle.go

```diff
@@ -0,0 +1,236 @@
+package libgeos
+
+/*
+#cgo linux CFLAGS: -I/usr/include
+#cgo linux LDFLAGS: -L/usr/lib -lgeos_c
+#include "geos_c.h"
+#include <stdlib.h>
+#include <string.h>
+
+void sf_error_handler(const char *message, void *userdata) {
+	strncpy(userdata, message, 1024);
+}
+
+GEOSContextHandle_t sf_init(void *userdata) {
+	GEOSContextHandle_t ctx = GEOS_init_r();
+	GEOSContext_setErrorMessageHandler_r(ctx, sf_error_handler, userdata);
+	return ctx;
+}
+
+*/
+import "C"
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"strings"
+	"unsafe"
+
+	"github.com/peterstace/simplefeatures/geom"
+)
+
+// Handle is an opaque handle that can be used to invoke libgeos operations.
+// Instances are not threadsafe and thus should not be used in a concurrent
+// manner.
+type Handle struct {
+	context C.GEOSContextHandle_t
+	errBuf  [1024]byte
+}
+
+// NewHandle creates a new libgeos handle.
+func NewHandle() (*Handle, error) {
+	h := &Handle{}
+	h.context = C.sf_init(unsafe.Pointer(&h.errBuf))
+	if h.context == nil {
+		return nil, errors.New("could not create libgeos context")
+	}
+	return h, nil
+}
+
+// err gets the last error message reported by libgeos as an error type. It
+// always returns a non-nil error. If no error message has been reported, then
+// it returns a generic error message.
+func (h *Handle) err() error {
+	msg := h.errMsg()
+	if msg == "" {
+		// No error stored, which indicates that the error handler didn't get
+		// trigged. The best we can do is give a generic error.
+		msg = "libgeos internal error"
```

Comment:
Just wondering why does it always return error?

Reply:
The `err()` function should only be called if something has gone wrong with an interaction with libgeos (as reported via the return code from the libgeos call). So we always want to give back some sort of error. But there could theoretically be a case where libgeos reports an error via its return value, but _didn't_ call the `sf_error_handler` callback (in which case we won't have an error message). Setting `msg` to a generic string is just so that the user gets some sort of error message (even though the message would not be very useful).